File: app/server/notify/channels/webhook.ts
/**
 * 自定义 Webhook 推送渠道
 */
import Logger from '../../utils/logger';
import { NotifyChannel, NotifyResult } from '../types';
import { httpClient } from '../httpClient';
import { hasConfig, getConfig } from '../config';

const logger = Logger.child({ channel: 'webhook' });

export const webhookChannel: NotifyChannel = {
    name: 'webhook',
    enabled: false,

    async send(text: string, desp: string): Promise<NotifyResult> {
        if (!hasConfig('WEBHOOK_URL')) {
            return { success: false, message: 'WEBHOOK_URL 未配置' };
        }

        const WEBHOOK_URL = getConfig('WEBHOOK_URL');
        const WEBHOOK_BODY = getConfig('WEBHOOK_BODY');
        const WEBHOOK_HEADERS = getConfig('WEBHOOK_HEADERS');
        const WEBHOOK_METHOD = getConfig('WEBHOOK_METHOD') || 'POST';
        const WEBHOOK_CONTENT_TYPE = getConfig('WEBHOOK_CONTENT_TYPE') || 'application/json';

        try {
            let body: string | undefined;
            const headers: Record<string, string> = {};

            // 设置 Content-Type
            headers['content-type'] = WEBHOOK_CONTENT_TYPE;

            // 解析自定义请求头
            if (WEBHOOK_HEADERS) {
                try {
                    const customHeaders = JSON.parse(WEBHOOK_HEADERS);
                    Object.assign(headers, customHeaders);
                } catch {
                    logger.warn('WEBHOOK_HEADERS 解析失败');
                }
            }

            // 构建请求体
            if (WEBHOOK_BODY) {
                body = WEBHOOK_BODY
                    .replace(/\$\{title\}/g, text)
                    .replace(/\$\{content\}/g, desp);
            } else {
                body = JSON.stringify({ title: text, content: desp });
            }

            const response = await httpClient.request(WEBHOOK_URL || '', {
                method: WEBHOOK_METHOD as 'GET' | 'POST' | 'PUT' | 'DELETE',
                headers,
                body
            });

            if (response.statusCode >= 200 && response.statusCode < 300) {
                logger.info({ statusCode: response.statusCode }, 'Webhook 发送成功');
                return { success: true, message: '发送成功' };
            } else {
                logger.warn({ statusCode: response.statusCode, body: response.body }, 'Webhook 发送失败');
                return { success: false, message: `HTTP ${response.statusCode}` };
            }
        } catch (err) {
            logger.error({ err }, 'Webhook 发送失败');
            return { success: false, error: err as Error };
        }
    }
};

// 检查是否启用
if (hasConfig('WEBHOOK_URL')) {
    webhookChannel.enabled = true;
}

export default webhookChannel;

File: app/server/internal/notify/channels/webhook.go
package channels

import (
	"encoding/json"
	"log/slog"
	"strings"

	"github.com/sushazhi/fnos-logmanager/internal/notify"
)

type Webhook struct {
	enabled bool
}

func (c *Webhook) Name() string {
	return "webhook"
}

func (c *Webhook) Enabled() bool {
	return c.enabled
}

func (c *Webhook) SetEnabled(enabled bool) {
	c.enabled = enabled
}

func (c *Webhook) Send(text, desp string) notify.NotifyResult {
	if !notify.HasConfig("WEBHOOK_URL") {
		return notify.NotifyResult{Success: false, Message: "WEBHOOK_URL 未配置"}
	}

	webhookURL := notify.GetConfig("WEBHOOK_URL")
	webhookBody := notify.GetConfig("WEBHOOK_BODY")
	webhookHeaders := notify.GetConfig("WEBHOOK_HEADERS")
	webhookMethod := notify.GetConfig("WEBHOOK_METHOD")
	if webhookMethod == "" {
		webhookMethod = "POST"
	}
	webhookContentType := notify.GetConfig("WEBHOOK_CONTENT_TYPE")
	if webhookContentType == "" {
		webhookContentType = "application/json"
	}

	headers := map[string]string{
		"content-type": webhookContentType,
	}

	// Parse custom headers from JSON
	if webhookHeaders != "" {
		var customHeaders map[string]string
		if err := json.Unmarshal([]byte(webhookHeaders), &customHeaders); err != nil {
			slog.Warn("WEBHOOK_HEADERS 解析失败", "error", err)
		} else {
			for k, v := range customHeaders {
				headers[k] = v
			}
		}
	}

	// Build body with template variables
	var body string
	if webhookBody != "" {
		body = strings.ReplaceAll(webhookBody, "${title}", text)
		body = strings.ReplaceAll(body, "${content}", desp)
	} else {
		b, _ := json.Marshal(map[string]string{
			"title":   text,
			"content": desp,
		})
		body = string(b)
	}

	resp, err := notify.HTTPRequest(webhookURL, notify.HttpRequestOptions{
		Method:  webhookMethod,
		Headers: headers,
		Body:    body,
	})
	if err != nil {
		slog.Error("Webhook 发送失败", "error", err)
		return notify.NotifyResult{Success: false, Message: err.Error(), Err: err}
	}

	if resp.StatusCode >= 200 && resp.StatusCode < 300 {
		slog.Info("Webhook 发送成功", "statusCode", resp.StatusCode)
		return notify.NotifyResult{Success: true, Message: "发送成功"}
	}

	slog.Warn("Webhook 发送失败", "statusCode", resp.StatusCode, "body", resp.Body)
	return notify.NotifyResult{Success: false, Message: "HTTP " + string(rune(resp.StatusCode))}
}

func init() {
	ch := &Webhook{}
	if notify.HasConfig("WEBHOOK_URL") {
		ch.SetEnabled(true)
	}
	notify.Registry.Register(ch)
}
